zbx_nginx_stats.py

```diff
@@ -38,81 +57,106 @@ def __repr__(self):
             return 'Metric(%r, %r, %r)' % (self.host, self.key, self.value)
         return 'Metric(%r, %r, %r, %r)' % (self.host, self.key, self.value, self.clock)
 
-def send_to_zabbix(metrics, zabbix_host='127.0.0.1', zabbix_port=10051):
-    
+def send_to_zabbix(metrics, zabbix_host, zabbix_port):
     j = json.dumps
     metrics_data = []
     for m in metrics:
         clock = m.clock or ('%d' % time.time())
-        metrics_data.append(('{"host":%s,"key":%s,"value":%s,"clock":%s}') % (j(m.host), j(m.key), j(m.value), j(clock)))
-    json_data = ('{"request":"sender data","data":[%s]}') % (','.join(metrics_data))
-    data_len = struct.pack('<Q', len(json_data))
-    packet = 'ZBXD\x01'+ data_len + json_data
+        #print('%s, %s, %s, %s' % (m.host, m.key, m.value, clock))
+        metrics_data.append('{"host":%s,"key":%s,"value":%s,"clock":%s}' % (j(m.host), j(m.key), j(m.value), j(clock)))
     
-    #print packet
-    #print ':'.join(x.encode('hex') for x in packet)
+    # Zabbix 3.0 protocol - https://www.zabbix.com/documentation/3.0/manual/appendix/items/activepassive
+    data = ('{"request":"sender data","data":[%s]}' % (','.join(metrics_data)))
+    data_length = len(data)
+    data_header = struct.pack('<Q', data_length) # 8 bytes
+    data_to_send = b'ZBXD\x01' + data_header + data.encode('utf-8')
+    #print(repr(data_to_send))
 
     try:
         zabbix = socket.socket()
-        zabbix.connect((zabbix_host, zabbix_port))
-        zabbix.sendall(packet)
-        resp_hdr = _recv_all(zabbix, 13)
-        if not resp_hdr.startswith('ZBXD\x01') or len(resp_hdr) != 13:
-            print 'Wrong zabbix response'
+        sock = zabbix
+        if args.use_ssl:
+            import ssl
+            context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+            context.options |= ssl.OP_NO_SSLv2
+            context.options |= ssl.OP_NO_SSLv3
+            context.verify_mode = ssl.CERT_REQUIRED
+            context.load_verify_locations(cafile=args.ssl_cafile)
+            context.load_cert_chain(certfile=args.ssl_certfile, keyfile=args.ssl_keyfile)
+
+            sock = context.wrap_socket(zabbix, server_hostname=zabbix_host)
```

Comment:
Correct. SSL using certificates - I did try to support TLS using PSK (pre-shared key) as well, but the ssl module doesn't support PSK, and I'd need a custom extension, which would introduce extra steps and requisites, so I avoided touching this.